### ISSUE TITLE
Verified stacker.news contributors

### ISSUE THREAD

**Issue by SatsAllDay** [stackernews/stacker.news#471]
# Verified stacker.news contributors

This feature proposal originated from a comment made on the most recent SNL episode, where one of my posts was mentioned and @huumn spoke about my recent contributions to the open source project, and Car pulled up my profile and there was discussion of having some kind of "contributor" indicator on my SN profile page.

The idea is to display a `verified stacker.news contributor` decoration on user profile, when applicable.

## How it works

* Users must link their associated GitHub account to their SN account as an auth method. This must be the GitHub account that has committed to the stacker.news repo.

* A new user setting boolean flag will be added, allowing the user to opt-in to showing this verified contributor status, if applicable. This will be configurable in the privacy section of user settings.

* A new field on user GraphQL API response payload -> `isVerifiedContributor: Boolean!`

* In the API resolver layer, we will query:
```bash
curl -L \
  -H "Accept: application/vnd.github+json" \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  https://api.github.com/repos/stackernews/stacker.news/contributors
```
[reference](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#list-repository-contributors)

* This API gives a paginated list of contributors of the repo. Load 100 at a time (max), load all pages until no more, and keep a map of id:username in memory.

* Cache this total set of contributors for an hour so as to reduce network calls (GitHub API w/o access token has strict rate limitations per hour)

When a user is requested via GraphQL API, we'll resolve the `isVerifiedContributor` field by checking:

1. Does the user have this setting enabled? No -> return false

2. Yes? Check if the user has a GitHub account (in the `accounts` table) linked? No -> return false

3. Yes? Check if the linked GitHub account ID exists in the current set of cached contributors? No -> return false

4. Yes? Return true

* The in-memory cache should be updated every hour, lazily on demand. This will cause one user query to take longer than others, but only once an hour.

* Only request this field for full user view (profile page)

* Display some kind of read-only decorator/icon on user profile pages based on the API response (maybe include the GitHub username so we can link to their contributions in GitHub or something?)

I want to emphasize that this would be opt-in only, and we are not persisting any additional information, nor are we requesting additional information from GitHub on the user's behalf. We're simply connecting the dots, if the user wants us to, to show stackers who has contributed to the project. Privacy is important!

Feedback? Is this something we'd like to have? I don't have an issue scrapping the idea if we don't collectively like it.

**Comment by ekzyis** [comment]
> Feedback? Is this something we'd like to have? I don't have an issue scrapping the idea if we don't collectively like it.

I like it!

However since this is probably just a one-time thing for every new contributor ... 

why not just add a `isContributor` column to the users table next to the `hideIsContributor` (or similar) and set it (more or less) manually?

Then it would literally just be a switch admins could enable for a user on SN instead of all of this:

> When a user is requested via GraphQL API, we'll resolve the isVerifiedContributor field by checking:
> 1. Does the user have this setting enabled? No -> return false
> 2. Yes? Check if the user has a GitHub account (in the accounts table) linked? No -> return false
> 3. Yes? Check if the linked GitHub account ID exists in the current set of cached contributors? No -> return false
> 4. Yes? Return true
> - The in-memory cache should be updated every hour, lazily on demand. This will cause one user query to take longer than others, but only once an hour.
> - Only request this field for full user view (profile page)
> - Display some kind of read-only decorator/icon on user profile pages based on the API response (maybe include the GitHub username so we can link to their contributions in GitHub or something?)

and this:

> Users must link their associated GitHub account to their SN account as an auth method. This must be the GitHub account that has committed to the stacker.news repo.

Requiring that you link your GitHub account to your SN profile may also not be ideal from a privacy perspective. I haven't looked what Github does with that information but I am not a fan of SSO - at least for personal use.

This could also just be a script that we run when there was a new contributor and they tell us their nym on SN.

Possibly, we don't even need a setting then. However, I think a setting is still good so they can choose to hide it at anytime.

I would like to add that I think such a setting should only be shown to contributors. So normal users wouldn't even see this setting.

**Comment by SatsAllDay** [comment]
> I like it!

Cool!

> why not just add a isContributor column to the users table next to the hideIsContributor (or similar) and set it (more or less) manually?

I was trying to avoid any manual intervention 

> Then it would literally just be a switch admins could enable for a user on SN instead of all of this:

I agree that we could implement it with far less code by making it a manually-set flag via admin operation.

> Requiring that you link your GitHub account to your SN profile may also not be ideal from a privacy perspective. I haven't looked what Github does with that information but I am not a fan of SSO - at least for personal use.

That's a fair point. I viewed it as something you could choose to do if you wanted to, so you had the ability to decide if this was too invasive, from a privacy perspective. But it would be nice to have alternate ways of verifying your contributor status aside from linking your GitHub account to SN. Again, I was trying to avoid manual intervention.

> Possibly, we don't even need a setting then. However, I think a setting is still good so they can choose to hide it at anytime.

I agree, I think we should have the setting to provide the option to hide it at any time.

> I would like to add that I think such a setting should only be shown to contributors. So normal users wouldn't even see this setting.

I agree with this, too.

**Comment by ekzyis** [comment]
> I was trying to avoid any manual intervention

Makes sense. However, this "manual intervention" could also not be that manual: A script could use the logic you described during deployment and update the appropriate user rows. Contributors could even add their nyms themselves in PRs if we use a file which maps their nyms / user ids to their github accounts which is then read to find the rows which need an update. :)

**Comment by SatsAllDay** [comment]
Cool, I like that idea!
